### ISSUE TITLE
open apk silently fails - missing FLAG_ACTIVITY_CLEAR_TOP

### ISSUE THREAD

**Issue by deeeed** [pwlin/cordova-plugin-file-opener2#163]
Hi,

I have been trying to implement an auto update function in my app.
I download an external apk to the external storage, then launch it with fileopener2 plugin.
Everything seems to work fine until the end of the installation which silently fails (it goes back to my original app without installing the new package).
![image](https://user-images.githubusercontent.com/4800095/29953270-549f2f86-8f01-11e7-8f30-5e0bb87f2004.png)

If I open the apk directly from the filesystem with a regular file explorer, the apk installs without any problem.
Here is the code I use to do the auto-update:

```
            let onSuccess = (entry) => {
                let fileURL = entry.toURL();
                console.debug("download complete!", fileURL);

                cordova.plugins.fileOpener2.open(
                    fileURL,
                    'application/vnd.android.package-archive',
                    {
                        error: function (e) {
                            console.log('Error status: ' + e.status + ' - Error message: ' + e.message);
                            _this._updating.set(false);
                        },
                        success: function () {
                            console.log('file opened successfully');
                            _this._updating.set(false);
                        }
                    }
                );
            }
```

the onSuccess method is called after the download has completed, and fileURL is in the form of **file:///storage/emulated/0/Download/newapp.apk**

Any idea what could be preventing the installation with fileOpener2? 
It could be a permission issue, but I thought the plugins would handle this.

**Comment by deeeed** [comment]
Just to answer myself, I checked the code of the plugin and noticed that
```
if((Build.VERSION.SDK_INT >= 23 && !contentType.equals("application/vnd.android.package-archive")) || ((Build.VERSION.SDK_INT == 24 || Build.VERSION.SDK_INT == 25) && contentType.equals("application/vnd.android.package-archive"))) {

					Context context = cordova.getActivity().getApplicationContext();
					path = FileProvider.getUriForFile(context, cordova.getActivity().getPackageName() + ".opener.provider", file);
					intent.setDataAndType(path, contentType);
					intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
					intent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
					intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP); // re-enable to compare installation

				 	List<ResolveInfo> infoList = context.getPackageManager().queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
					for (ResolveInfo resolveInfo : infoList) {
				    String packageName = resolveInfo.activityInfo.packageName;
				    context.grantUriPermission(packageName, path, Intent.FLAG_GRANT_WRITE_URI_PERMISSION | Intent.FLAG_GRANT_READ_URI_PERMISSION);
					}
				}
				else {
					intent.setDataAndType(path, contentType);
					intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
				}
```
I uncommented the line 
```
intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
```
Now my installation completes but the newly installed apk doesn't launch automatically.
Is there a flag I could use to make it run automatically after it has been installed?

**Comment by pallaviMN** [comment]
Hi @deeeed , Did u get any solution to this? Are you using release APK or debug APK. For me while using debug APK, it works first time but second time it says "Error Parsing Package".

**Comment by deeeed** [comment]
Hi @pallaviMN, actually I thought it was solved and then also noticed that on the second time it doesn't work.
I get an error saying the apk is corrupted... (but it's working fine if I install the downloaded apk outside the app)

**Comment by pallaviMN** [comment]
@deeeed , yes even for me the similar issue. If i navigate to file explorer & try to install it works fine

**Comment by shnist** [comment]
hi @deeeed thanks for raising this issue. Were you able to resolve it in the end? There was a significant refactor around this functionality, which was released recently from version `2.1.0`. If you try this version it may now work.

**Comment by shnist** [comment]
going to close this issue for now due to inactivity. Feel free to re-open if the issue persists.
